ethexe/validator/src/lib.rs

```diff
@@ -16,99 +16,438 @@
+    router_address: Address,
 }
 
 pub struct Config {
     pub pub_key: PublicKey,
     pub router_address: Address,
 }
 
-pub struct Validator {
-    pub_key: PublicKey,
-    signer: Signer,
-    current_codes: Vec<CodeCommitment>,
-    current_blocks: Vec<BlockCommitment>,
-    router_address: Address,
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct BlockCommitmentValidationRequest {
+    pub block_hash: H256,
+    pub prev_commitment_hash: H256,
+    pub pred_block_hash: H256,
+    pub transitions_digest: Digest,
+}
+
+impl From<&BlockCommitment> for BlockCommitmentValidationRequest {
+    fn from(commitment: &BlockCommitment) -> Self {
+        Self {
+            block_hash: commitment.block_hash,
+            prev_commitment_hash: commitment.prev_commitment_hash,
+            pred_block_hash: commitment.pred_block_hash,
+            transitions_digest: commitment.transitions.as_digest(),
+        }
+    }
+}
+
+impl AsDigest for BlockCommitmentValidationRequest {
+    fn as_digest(&self) -> Digest {
+        let mut message = Vec::with_capacity(3 * size_of::<H256>() + size_of::<Digest>());
+
+        message.extend_from_slice(self.block_hash.as_bytes());
+        message.extend_from_slice(self.prev_commitment_hash.as_bytes());
+        message.extend_from_slice(self.pred_block_hash.as_bytes());
+        message.extend_from_slice(self.transitions_digest.as_ref());
+
+        message.as_digest()
+    }
 }
 
 impl Validator {
     pub fn new(config: &Config, signer: Signer) -> Self {
         Self {
             signer,
             pub_key: config.pub_key,
-            current_codes: vec![],
-            current_blocks: vec![],
             router_address: config.router_address,
         }
     }
 
-    pub fn has_codes_commit(&self) -> bool {
-        !self.current_codes.is_empty()
-    }
-
-    pub fn has_transitions_commit(&self) -> bool {
-        !self.current_blocks.is_empty()
-    }
-
     pub fn pub_key(&self) -> PublicKey {
         self.pub_key
     }
 
-    pub fn codes_aggregation(&mut self) -> Result<AggregatedCommitments<CodeCommitment>> {
+    pub fn address(&self) -> Address {
+        self.pub_key.to_address()
+    }
+
+    pub fn aggregate<C: AsDigest>(&self, commitments: Vec<C>) -> Result<AggregatedCommitments<C>> {
         AggregatedCommitments::aggregate_commitments(
-            self.current_codes.clone(),
+            commitments,
             &self.signer,
             self.pub_key,
             self.router_address,
         )
     }
 
-    pub fn blocks_aggregation(&mut self) -> Result<AggregatedCommitments<BlockCommitment>> {
-        AggregatedCommitments::aggregate_commitments(
-            self.current_blocks.clone(),
+    pub fn validate_code_commitments(
+        &mut self,
+        db: &impl CodesStorage,
+        requests: impl IntoIterator<Item = CodeCommitment>,
+    ) -> Result<(Digest, Signature)> {
+        let mut commitment_digests = Vec::new();
+        for request in requests.into_iter() {
```

Comment:
```suggestion
        for request in requests {
```
Conversion is unnecessary because it's automatic

Reply:
Done